### ISSUE TITLE
tf.keras.initializers.RandomNormal with the same seed produces different tensors in TF27 vs older TF2 versions

### ISSUE THREAD

**Issue by q-ycong-p** [keras-team/keras#15899]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux x86_64
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): `pip install tensorflow==2.7.0`
- Python version: 3.8.12
- Bazel version (if compiling from source): N/A
- GPU model and memory: N/A
- Exact command to reproduce: https://gist.github.com/q-ycong-p/03a48b3974aead91758de9cc51be6d48

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

 TF27 introduces [breaking changes](https://github.com/tensorflow/tensorflow/releases) to `tf.keras.initializers` behavior. Initializers generated from the same seeds in TF27 are inconsistent against those from older TF2 versions. Specifically, older TF2 uses `tf.random.stateless_normal`([ref](https://github.com/keras-team/keras/blob/3a33d53ea4aca312c5ad650b4883d9bac608a32e/keras/initializers/initializers_v2.py#L960)) while  TF27 uses  `tf.random.Generator.from_seed` ([ref](https://github.com/keras-team/keras/blob/2c48a3b38b6b6139be2da501982fd2f61d7d48fe/keras/backend.py#L1771)). This is problematic for users, because one cannot use the same user script to generate models consistent in TF27 and older TF2. For example, unit testing against pre-saved TF26 models is impacted.

**Describe the current behavior**.
Initializers generated given the same seed give **different** values in TF27 vs TF26. 
Try running [this script](https://gist.github.com/q-ycong-p/03a48b3974aead91758de9cc51be6d48) in TF27 vs TF26:
<img width="450" alt="Screen Shot 2022-01-13 at 2 44 13 PM" src="https://user-images.githubusercontent.com/50014064/149420783-dce08a36-886d-4f10-8b83-725fe2bccbcc.png">
<img width="444" alt="Screen Shot 2022-01-13 at 2 44 08 PM" src="https://user-images.githubusercontent.com/50014064/149420790-7594df8e-290f-4f25-8b91-e8637cfe16db.png">

**Note:** I've read [#15586](https://github.com/keras-team/keras/issues/15586), and tried suggested tweaks in shared gist. Note that the bug's root cause *might* be same, but it's not the same issue: here the issue is inconsistency between TF versions, not between calls/sessions.

**Describe the expected behavior**.
Initializers generated given the same seed give **same** values in TF27 vs TF26. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.
gist snapshot: https://gist.github.com/q-ycong-p/03a48b3974aead91758de9cc51be6d48. Pls compare the results in TF27 vs TF26.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in tf v2.5,v2.7 and nightly.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/2b3ddf5032904e064bbba04e3d1b27aa/untitled191.ipynb).

**Comment by qlzh727** [comment]
@wangpengmit here who is the owner of RNG behavior in TF. I don't think same seed will ensure to produce exact same value across different version of TF binary is part of the public TF API.

**Comment by q-ycong-p** [comment]
> @wangpengmit here who is the owner of RNG behavior in TF. I don't think same seed will ensure to produce exact same value across different version of TF binary is part of the public TF API.

I see. But intuitively, a user would expect initializers to be the same from same seed, after a change in TF2 version right? It might not be crucial for training production model, but it'd be an expected behavior when users design unit testing and alike.

**Comment by wangpengmit** [comment]
As @qlzh727 said, different versions of TF don't guarantee generating the same random numbers. See [here](https://www.tensorflow.org/guide/random_numbers) and [here](https://www.tensorflow.org/guide/versions#what_is_not_covered). As a best practice, unit tests shouldn't depend on specific random numbers.

**Comment by q-ycong-p** [comment]
> As @qlzh727 said, different versions of TF don't guarantee generating the same random numbers. See [here](https://www.tensorflow.org/guide/random_numbers) and [here](https://www.tensorflow.org/guide/versions#what_is_not_covered). As a best practice, unit tests shouldn't depend on specific random numbers.

Got it. My use case of unit test was to compare converted models against a correct model generated by an older TF2 version, leading to the original issue. Thanks for linking to the API doc, I now understand it's not a covered behavior. But can we make it a feature request, since the support of consistent RNG based on seeds would make it easier for user?

**Comment by wangpengmit** [comment]
We intentionally reserved the right to change RNG results across TF versions because any internal change to RNGs will likely change their results, hence preserving RNG results across TF versions basically means freezing RNG development, which is unacceptable. For you use case, I suggest you (at the beginning of training) manually set the old and new models' trainable parameters to the same values, to bypass RNG differences.

**Comment by q-ycong-p** [comment]
> We intentionally reserved the right to change RNG results across TF versions because any internal change to RNGs will likely change their results, hence preserving RNG results across TF versions basically means freezing RNG development, which is unacceptable. For you use case, I suggest you (at the beginning of training) manually set the old and new models' trainable parameters to the same values, to bypass RNG differences.

Understandable. Thank you!

**Comment by sachinprasadhs** [comment]
@q-ycong-p If your issue is addressed, could you please move this issue to closed. Thanks!

**Comment by q-ycong-p** [comment]
Closing the issue since it's reasonable to not support consistent RNG behavior from development point of view. Thank you for all responded.
